### PR TITLE
V6 - ShallowEqual need nextState

### PR DESCRIPTION
**Pull request by Darmody**
Because `shallowequal(undefined, null) === false`, should add mssing `nextState` param to `shallowCompare`

I believe some rerender cases are caused by this problem.

### CODE REVIEW

**Comment by erikras** [comment]
That was intentionally left out because these components are stateless, and therefore `this.state` and `nextState` will always be `undefined`. Have you seen a case where one is `null` and the other is `undefined`?

**Comment by Darmody** [comment]
@erikras Yep, I check it in my project, `this.state` is null.And I think null is the default value of `this.state`?

File: src/__tests__/Field.spec.js [hunk 1]
```diff
@@ -606,61 +606,61 @@ const describeField = (name, structure, combineReducers, expect) => {
     // ----------------------------------------------
     // Uncomment this to confirm that #1024 is fixed.
     // ----------------------------------------------
```
Comment (erikras) [inline src/__tests__/Field.spec.js]: Why did you comment this out again?

Comment (Darmody) [inline src/__tests__/Field.spec.js]: Because it failed, I think maybe it has not been fixed correctly.

Comment (Darmody) [inline src/__tests__/Field.spec.js]: @erikras Oh, I think that syncError is not related to props, so when I avoid the unnecessary rerender, the test failed.

File: src/Field.js [hunk 2]
```diff
@@ -30,6 +31,11 @@ const createField = ({ deepEqual, getIn, setIn }) => {
         this.ConnectedField =
           createConnectedField(this.context._reduxForm, { deepEqual, getIn }, nextProps.name)
       }
+
+      const syncError = this.getSyncError()
+      if (syncError !== this.state.syncError) {
+        this.setState({ syncError })
+      }
```
Comment (Darmody) [inline src/Field.js]: @erikras Is it reasonable?

Comment (erikras) [inline src/Field.js]: No. The results of `getSyncError()` may change over time.

Comment (Darmody) [inline src/Field.js]: I think it is a problem that can not track error properly. 
If field`s validation depend on itself, it is easy to track the error.
If field's validation depend on other field, there is no way to track the error unless make component  rerender every time.

Comment (Darmody) [inline src/Field.js]: https://github.com/erikras/redux-form/pull/1120

this PR works because `Field` component always rerender.

File: src/Field.js [hunk 3]
```diff
@@ -16,8 +16,11 @@ const createField = ({ deepEqual, getIn, setIn }) => {
       this.normalize = this.normalize.bind(this)
     }
 
-    shouldComponentUpdate(nextProps) {
-      return shallowCompare(this, nextProps)
+    shouldComponentUpdate(nextProps, nextState, nextContext) {
+      if (this.context._reduxForm.syncErrors !== nextContext._reduxForm.syncErrors) {
```
Comment (Darmody) [inline src/Field.js]: I believe it works!

Comment (Darmody) [inline src/Field.js]: Oops, it may not...

**Comment by Darmody** [comment]
I apply the commit 4ad06b59041d17144e3e63e01de78cf0230f1267 to tag v6-alpha.15, all validation work fine in my project.

But when I apply it to branch v6, validation broken.

Since I don't know how to fix the problem, closed the PR.

**Comment by erikras** [comment]
I have concluded that you are correct about needing to pass along `nextState`. But that _does_ cause the test that you commented out to fail, although I am not sure why.

**Comment by Darmody** [comment]
@erikras Because  error states are stored in context, whenever error changed, the Field component will not rerender. That makes the test failed.

But I don't know how to fix it..
